File: ios/MenCare/Core/Network/Dto/BookingDto.swift
import Foundation

// Agenda e lista d'attesa — docs/API.md §Agenda e §Lista d'attesa.

struct AppointmentDto: Decodable {
    let id: String
    let clientId: String
    let operatorId: String
    let serviceIds: [String]
    let date: LocalDate
    let time: LocalTime
    let startsAt: String
    let durationMinutes: Int
    let totalPriceCents: Int64
    let status: String
    let channel: String
    let noteForOperator: String?
    let cancelledBy: String?

    func toDomain() -> Appointment {
        Appointment(
            id: id,
            clientId: clientId,
            operatorId: operatorId,
            serviceIds: serviceIds,
            // `date` + `time` sono l'ora da muro del salone: e' quella che le
            // schermate mostrano. `startsAt` serve solo al server.
            start: date.atTime(time),
            durationMinutes: durationMinutes,
            totalPriceCents: totalPriceCents,
            status: WireEnum.appointmentStatus(status),
            channel: WireEnum.channel(channel),
            noteForOperator: noteForOperator,
            cancelledBy: WireEnum.cancellationActor(cancelledBy)
        )
    }
}

struct AppointmentsResponseDto: Decodable {
    let appointments: [AppointmentDto]
}

/// `GET /booking/availability` — gli slot li calcola il server, preavviso di
/// 30 minuti e unione "qualsiasi operatore" compresi.
struct AvailabilityDto: Decodable {
    let date: LocalDate
    let slots: [LocalTime]
}

/// `GET /booking/days` — giorni con almeno uno slot e giorni pieni (aperti ma
/// senza piu' niente di libero: quelli offrono la lista d'attesa).
struct AvailabilityRangeDto: Decodable {
    let available: [LocalDate]
    let fullyBooked: [LocalDate]
}

/// `GET /booking/next-availability` — prima disponibilita' di ogni operatore.
/// Mappata ma non usata: vedi la nota su `BookingEndpoint.nextAvailability`.
struct NextAvailabilityDto: Decodable {
    struct Entry: Decodable {
        let operatorId: String
        let date: LocalDate
        let time: LocalTime
    }

    let operators: [Entry]
}

struct WaitlistEntryDto: Decodable {
    let id: String
    let clientId: String
    let date: LocalDate
    let time: LocalTime?
    let operatorId: String?
    let serviceIds: [String]
    let durationMinutes: Int
    let totalPriceCents: Int64
    let position: Int
    let status: String

    func toDomain() -> WaitlistEntry {
        WaitlistEntry(
            id: id,
            clientId: clientId,
            date: date,
            time: time,
            operatorId: operatorId,
            serviceIds: serviceIds,
            durationMinutes: durationMinutes,
            totalPriceCents: totalPriceCents,
            // La posizione in coda la calcola il server: era una delle cose che
            // le due app facevano da sole, ognuna a modo suo.
            position: position,
            status: WireEnum.waitlistStatus(status)
        )
    }
}

struct WaitlistResponseDto: Decodable {
    let entries: [WaitlistEntryDto]
}

// MARK: - Corpi delle scritture

struct BookRequestDto: Encodable {
    let clientId: String?
    let operatorId: String?
    let serviceIds: [String]
    let date: LocalDate
    let time: LocalTime
    let noteForOperator: String?
    let channel: String?
    /// "Modifica": l'appuntamento che questo sostituisce (vedi `BookingRequest`).
    let replacesAppointmentId: String?
}

struct RescheduleRequestDto: Encodable {
    let date: LocalDate
    let time: LocalTime
    let operatorId: String?
}

struct StatusRequestDto: Encodable {
    let status: String
}

struct JoinWaitlistRequestDto: Encodable {
    let date: LocalDate
    let time: LocalTime?
    let operatorId: String?
    let serviceIds: [String]
}

File: android/core/network/src/main/kotlin/com/devora/mencare/core/network/dto/BookingDto.kt
package com.devora.mencare.core.network.dto

import kotlinx.serialization.Serializable

/**
 * Agenda, disponibilità e lista d'attesa: `/v1/booking/…`, `/v1/appointments…`
 * e `/v1/waitlist*`.
 *
 * `date` e `time` sono l'orario da muro del salone (Europe/Rome) ed è quello
 * che l'app mostra; `startsAt` è lo stesso momento in ISO completo e serve solo
 * a chi deve ordinare fra fusi diversi.
 */

@Serializable
data class AppointmentDto(
    val id: String,
    val clientId: String,
    val operatorId: String,
    val serviceIds: List<String> = emptyList(),
    val date: String,
    val time: String,
    val startsAt: String? = null,
    val durationMinutes: Int,
    /** Assente nelle risposte al ruolo STAFF: gli importi non gli vengono proprio selezionati. */
    val totalPriceCents: Long? = null,
    val status: String,
    val channel: String = "APP",
    val noteForOperator: String? = null,
    val cancelledBy: String? = null,
)

@Serializable
data class AppointmentsEnvelope(val appointments: List<AppointmentDto> = emptyList())

@Serializable
data class AvailabilityDto(val date: String, val slots: List<String> = emptyList())

/** `/booking/days`: giorni con almeno uno slot libero e giorni aperti ma pieni. */
@Serializable
data class AvailableDaysDto(
    val available: List<String> = emptyList(),
    val fullyBooked: List<String> = emptyList(),
)

@Serializable
data class NextAvailabilityDto(val operatorId: String, val date: String, val time: String)

@Serializable
data class NextAvailabilityEnvelope(val operators: List<NextAvailabilityDto> = emptyList())

@Serializable
data class CreateAppointmentBody(
    val clientId: String? = null,
    /** Assente = "qualsiasi operatore": la scelta la fa il server. */
    val operatorId: String? = null,
    val serviceIds: List<String>,
    val date: String,
    val time: String,
    val noteForOperator: String? = null,
    val channel: String? = null,
    /** "Modifica": l'appuntamento che questo sostituisce, nella stessa transazione. */
    val replacesAppointmentId: String? = null,
)

@Serializable
data class RescheduleBody(val date: String, val time: String, val operatorId: String? = null)

@Serializable
data class StatusBody(val status: String)

@Serializable
data class WaitlistEntryDto(
    val id: String,
    val clientId: String,
    val date: String,
    /** `null` = tutta la giornata. */
    val time: String? = null,
    val operatorId: String? = null,
    val serviceIds: List<String> = emptyList(),
    val durationMinutes: Int = 0,
    val totalPriceCents: Long = 0,
    /** Posizione in coda: la calcola il server, l'app non la ricostruisce. */
    val position: Int = 0,
    val status: String = "WAITING",
)

@Serializable
data class WaitlistEnvelope(val entries: List<WaitlistEntryDto> = emptyList())

@Serializable
data class JoinWaitlistBody(
    val date: String,
    val time: String? = null,
    val operatorId: String? = null,
    val serviceIds: List<String>,
)
